mlem/contrib/sagemaker/meta.py

```diff
@@ -0,0 +1,484 @@
+    def _call_method(self, name, args):
+        return self.get_predictor().predict(args)
+
+
+class SagemakerDeployState(DeployState):
+    """State of SageMaker deployment"""
+
+    type: ClassVar = "sagemaker"
+
+    image: Optional[DockerImage] = None
+    """Built image"""
+    image_tag: Optional[str] = None
+    """Built image tag"""
+    model_location: Optional[str] = None
+    """Location of uploaded model"""
+    endpoint_name: Optional[str] = None
+    """Name of SageMaker endpoint"""
+    endpoint_model_hash: Optional[str] = None
+    """Hash of deployed model"""
+    method_signature: Optional[Signature] = None
+    """Signature of deployed method"""
+    region: Optional[str] = None
+    """AWS Region"""
+    previous: Optional["SagemakerDeployState"] = None
+    """Previous state"""
+
+    @property
+    def image_uri(self):
+        if self.image is None:
+            if self.image_tag is None:
+                raise ValueError(
+                    "Cannot get image_uri: image not built or not specified prebuilt image uri"
+                )
+            return self.image_tag
+        return self.image.uri
+
+    def get_predictor(self, session: sagemaker.Session):
+        predictor = sagemaker.Predictor(
+            endpoint_name=self.endpoint_name,
+            sagemaker_session=session,
+            serializer=JSONSerializer(),
+            deserializer=JSONDeserializer(),
+        )
+        return predictor
+
+
+class SagemakerDeployment(MlemDeployment):
+    """SageMaker Deployment"""
+
+    type: ClassVar = "sagemaker"
+    state_type: ClassVar = SagemakerDeployState
+
+    method: str = "predict"
+    """Model method to be deployed"""
+    image_tag: Optional[str] = None
+    """Name of the docker image to use"""
+    use_prebuilt: bool = False
+    """Use pre-built docker image. If True, image_name should be set"""
+    model_arch_location: Optional[str] = None
+    """Path on s3 to store model archive (excluding bucket)"""
+    model_name: Optional[str]
+    """Name for SageMaker Model"""
+    endpoint_name: Optional[str] = None
+    """Name for SageMaker Endpoint"""
+    initial_instance_count: int = 1
+    """Initial instance count for Endpoint"""
+    instance_type: str = "ml.t2.medium"
+    """Instance type for Endpoint"""
+    accelerator_type: Optional[str] = None
+    "The size of the Elastic Inference (EI) instance to use"
+
+    @validator("use_prebuilt")
+    def ensure_image_name(  # pylint: disable=no-self-argument
+        cls, value, values  # noqa: B902
+    ):
+        if value and "image_name" not in values:
+            raise ValueError(
+                "image_name should be set if use_prebuilt is true"
+            )
+        return value
+
+    def _get_client(self, state: "SagemakerDeployState"):
+        return SagemakerClient(
+            endpoint_name=state.endpoint_name,
+            aws_vars=self.get_env().get_session_and_aws_vars(
+                region=state.region
+            )[1],
```

Comment:
why `[1]`? can we return a `dict` instead of a `tuple` here? I mean in `init_aws_vars`

Reply:
answered in other thread